### PR TITLE
Fix Union type with dataclass ambiguous error and support superset comparison

### PR DESCRIPTION
**Pull request by mao3267**
## Summary
We only support json schema draft 

1. same json schema
2. big -> small (superset)
3. inherit case (superset)

## Tracking issue
Related to #5489 

## Why are the changes needed?
When a function accepts a `Union` of two dataclasses as input, Flyte cannot distinguish which dataclass matches the user's input. This is because Flyte only compares the simple types, and both dataclasses are identified as `flyte.SimpleType_STRUCT` in this scenario. As a result, there will be multiple matches, causing ambiguity and leading to an error.

`union_test_dataclass.py`
```py
from typing import Union
from dataclasses import dataclass
from dataclasses_json import dataclass_json
from flytekit import task, workflow

@dataclass_json 
@dataclass
class A:
    a: int
    

@dataclass_json
@dataclass
class B:
    b: int


@task
def bar() -> A:
    return A(a=1)

@task
def foo(inp: Union[A, B]):
    print(inp)

@workflow
def wf():
    v = bar()
    foo(inp=v)
```

## What changes were proposed in this pull request?
1. To distinguish between different types using protobuf struct (dataclass, Pydantic.BaseModel), we compare their JSON schemas generated by `marshmallow_jsonschema.JSONSchema` (draft-07)  or `mashumaro.jsonschema.build_json_schema` (draft 2020-12) for dataclass and Pydantic.BaseModel for itself. To check equivalence, we compare the bytes from marshaling the json schemas if they are in the same draft version. For now, we only consider supporting the comparison of schemas with the same version. 
2. We plan to support superset matching for dataclass/Pydantic.BaseModel with schemas in draft 2020-12, meaning that class A and class supersetA can be a match in the following example: (Pydantic.BaseModel example is in the screenshot section)

3. Unit tests for different versions of json schema, including one-level, two-level, and strict subset examples in draft 2020-12 from mashumaro.

## How was this patch tested? (my local testing repo [here](https://github.com/mao3267/local_test))
1. Should Pass Tests
      - Run an example using union input with identical dataclass on remote (union_test_dataclass.py)
      - Run an example with inheritance on remote (union_test_inheritance.py)
      - Run an example using union input with identical BaseModel class on remote (union_test_basemodel.py)
2. Should Fail Tests
      - upstream parent and downstream inherit child (dataclass_child_to_parent.py)
      - two level dataclass with different properties (dataclass_two_level.py)

### Setup process
```
git clone https://github.com/flyteorg/flyte.git
gh pr checkout 5858
make compile
POD_NAMESPACE=flyte ./flyte start --config flyte-single-binary-local.yaml
```
### Screenshots
1. Example using union input with identical dataclass on remote (union_test_dataclass.py)
    
image

2. Example with inheritance on remote (union_test_inheritance.py)

image

3. Example using union input with identical BaseModel class on remote (union_test_basemodel.py)

image

4. Example with upstream parent and downstream inherit child (dataclass_child_to_parent.py)

image

5. Two level dataclass with different properties (dataclass_two_level.py)

image

### Optional values in JSON Schemas
While handling `Optional` values in Python, both `NoneType` and the target type are accepted. However, when defining such values, default values must still be provided. This is why `Optional` properties without default values are marked as `required` in JSON schemas.

### Ambiguity vs. Inheritance
Currently, we compare only the structure of the dataclass, excluding the title, as including it would block inheritance. This approach introduces ambiguity because it allows dataclasses with different names but the same structure to match. To resolve this, we may require additional information from Flytekit to determine whether two dataclasses have an inheritance relationship. The following example illustrates this challenge: B should match A, but C should not.

```py
@dataclass
class A:
    a: int

@dataclass
class B(A):
    b: int

@dataclass
class C:
    a: int
```

### Json Schema Compare Package (LGPL License)
We are now using [json-schem-compare](https://pkg.go.dev/gitlab.com/yvesf/json-schema-compare) package to do strict subset matching. However, since the underlying JSON Schema compiler only supports drafts up to draft-7, it does not handle the prefixItems field when tuples are used in a dataclass.

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
None

## Docs link
TODO

### CODE REVIEW

**Comment by Future-Outlier** [comment]
> One note @mao3267, this problem does not only affect dataclasses but any other type that uses protobuf struct as transport. Even combinations of different types that all use protobuf struct. We for instance have an internal type transformer for pydantic base models (historic reasons before an official one was introduced as a plugin). It has exactly the same problem because it also uses protobuf struct. It would be nice if the solution found for this problem was general and not only working for dataclasses. That's what I'm sligthy worried about when reading
> 
> > distinguish between different dataclasses, we compare their JSON schemas generated by either marshmallow_jsonschema.JSONSchema (draft-07) or mashumaro.jsonschema.build_json_schema
> 
> Maybe we can use the literal type's "type structure" field for this? Or we should document how transformers for other types can provide the schema in a way that they can "participate in the logic".
> 
> TL;DR: It would be nice if the compiler logic in flytepropeller didn't have "special treatment" for dataclasses but general treatment for json-like structures with schemas that the dataclass transformer makes use of - but other transformers can as well.

just dicussed with @mao3267 , he will explain how it works now, this will support both dataclass and pydantic basemodel in summary.

**Comment by Future-Outlier** [comment]
Let's get this done this week @mao3267

**Comment by mao3267** [comment]
> just dicussed with @mao3267 , he will explain how it works now, this will support both dataclass and pydantic basemodel in summary.

Currently, we support both `Pydantic BaseModel` and `dataclass`, including their combinations and nested structures. For `dataclass_json`, we only support equivalence without superset matching due to significant differences between its JSON schema (draft-07) and the newer version used by both Pydantic and dataclass (draft 2020-12).

Although the schemas from `Pydantic BaseModel` and `dataclass` adhere to the same version, they differ in how they handle fields. For instance, only the schema generated by `Pydantic BaseModel` records the title in properties. Additionally, the `required` and `additionalProperties` fields are omitted if no required properties exist or if additional properties are disallowed. To address these discrepancies, we preprocess the schema before comparison, which involves removing the title field, and modify the logic while comparing for the `required` and `additionalProperties` field.

**Comment by wild-endeavor** [comment]
the exact match part is okay and we should fix that but I'm confused about the other part.

looking at this example
```
@task
def my_task(input: Union[supersetA, B]):
    print(input)

@workflow
def wf():
    a = foo()
    my_task(a)
```

can you explain why this should work?  `foo` creates an `A` object with only one field.  `my_task` is a task that takes in `supersetA` or `B`.  `B` is not relevant here.  `supersetA` takes in three fields.

Why doesn't mypy complain in this example?  Or does it?

I almost feel like if we're going to go down this route, it should be the other way around.  If `foo` returned `supersetA` and `my_task` took in union of `A` and `B`.  The reason is

cc @fg91 if you want to take a look as well.

wasn't there another pr where we were discussing an LGPL library also?

File: flytepropeller/pkg/compiler/validators/typing.go [hunk 1]
```diff
@@ -19,6 +19,8 @@ type trivialChecker struct {
 }
 
 func removeTitleFieldFromProperties(schema map[string]*structpb.Value) {
+	// TODO: Explain why we need this
+	// TODO: givse me example about dataclass vs. Pydantic BaseModel
```
Comment (mao3267) [inline flytepropeller/pkg/compiler/validators/typing.go:23]: This is an example comparing dataclass and Pydantic BaseModel. As shown, the schema for dataclass includes a title field that records the name of the class. Additionally, the `additionalProperties` field is absent from the Pydantic BaseModel schema because its value is `false`.

| dataclass | Pydantic.BaseModel |
| - | - |
| image | image |

[the remark this answers is not in the document]
Comment (mao3267) [reply flytepropeller/pkg/compiler/validators/typing.go:23]: To add the comment, writing the entire schema would make it too lengthy. Would it be acceptable to use something like this instead?

```
class A:
	a: int

Pydantic.BaseModel: 	{"properties": {"a": {"title": "A", "type": "integer"}}}
dataclass: 			{"properties": {"a": {"type": "integer"}}, "additionalProperties": false}
```

[the remark this answers is not in the document]
Comment (fg91) [reply flytepropeller/pkg/compiler/validators/typing.go:23]: Are you proposing to preprocess the schemas so that one can mix and match dataclasses and base models given their schemas are aligned? I personally feel this is not necessary and think it would be totally acceptable to consider a dataclass and a base model *not* a match by default. Especially if this makes things a lot more complicated in the backend otherwise. What do you think about this?

**Comment by fg91** [comment]
> Currently, we support both Pydantic BaseModel and dataclass, including their combinations and nested structures. 

@Future-Outlier @mao3267 let's consider the scenario that an org wants to build a custom internal type transformer for a type similar to dataclasses or base models. Is there a way they can provide the schema of their type in the `to_literal_type` method of their type transformer so that the backend can automatically perform schema checks for Union types? Or are there implementation details that are required in the backend too that limit this to dataclasses/base models and would be required for any additional json-like type?

The latter would be slightly concerning to me.

It would be really great if there was a tutorial in https://docs.flyte.org/en/latest/api/flytekit/types.extend.html in the end that documents how users need to provide the schema in their respective `to_literal_type` implementation so that the backend can automatically make use of it in the union type checker.

File: flytepropeller/pkg/compiler/validators/typing.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mao3267) [reply flytepropeller/pkg/compiler/validators/typing.go:23]: I think it actually make things more complicated, will remove related logic.

**Comment by mao3267** [comment]
Replying @wild-endeavor 

> then this should not work correct? (because c is missing).

Yes. This should not work.

> Is there a comment somewhere that explains why we no longer need, or can't use, that library or a library like it?

No. I just decided to use another package that directly compares the JSON schema. This is not documented anywhere.


Replying @fg91 
> My question above wasn't specifically about base models but about how generalizable the solution is.
Let's consider the scenario that an org wants to build a custom internal type transformer for a json-like type similar to dataclasses or base models.
Is there a way they can provide the schema of their type in the to_literal_type method of their type transformer so that the backend can automatically perform schema checks for Union types? Or are there implementation details that are required in the backend too that limit this to dataclasses/base models and would be required for any additional json-like type?

> The latter would be slightly concerning to me. Glancing over the code gives me the impression that there is quite a bit of dataclass/pydantic logic we need to apply. I wonder whether this could be done in the respective flytekit type transformer so that the backend is agnostic to the respective type and as long as the type transformer provides the schema in the right way, the backend can make use of it.

> It would be really great if there was a tutorial in https://docs.flyte.org/en/latest/api/flytekit/types.extend.html in the end that documents how users need to provide the schema in their respective to_literal_type implementation so that the backend can automatically make use of it in the union type checker.

To support any other custom types, could we limit the provided schema from a certain package like Mashumaro for the compatibility check? Without limitations, it will be hard to cover all kinds of scenarios. Or does anyone know possible solutions to support JSON schema from different versions and packages?

**Comment by fg91** [comment]
> To support any other custom types, could we limit the provided schema from a certain package like Mashumaro for the compatibility check? Without limitations, it will be hard to cover all kinds of scenarios. Or does anyone know possible solutions to support JSON schema from different versions and packages?

Yes, I think restricting what kind of schema needs to be supplied is absolutely reasonable! I think it would be good to add a tutorial here https://docs.flyte.org/en/latest/api/flytekit/types.extend.html that states something like "if you want propeller to understand the schema of your type e.g. to distinguish in union types, you need to provide a schema in the to_literal_type method in this specific way". And I personally feel that the dataclass and pydantic type transformers should provide the schema in this general way so that the backend doesn't have to have type specific implementations for dataclasses/base models.
What do you think about this?

**Comment by wild-endeavor** [comment]
@fg91 and @EngHabu mind chiming in on dataclass compatibility?  Just thinking about it from the basics, if I have json schemas representing two dataclasses, let's say,

```python
@dataclass
class A2:
    a: int
    b: Optional[int] = None
    c: str = "hello"
```
and
```python
@dataclass
class A1:
    a: int
```

which of the following two are valid?
Case 1
```python
def wf():
  a1 = create_a1()  # -> A1
  use_a2(a2=a1)  # a2: A2
```

Case 2
```python
def wf():
  a2 = create_a2()  # -> A2
  use_a1(a1=a2)  # a1: A1
```

Just thinking about compatibility in the loosest sense, both should be valid.  The reason is that in the first case, when calling the downstream task `use_a2`, fields `b` and `c` have defaults.

In the second case, the `a` field can be taken from the `a2` object and `b` and `c` discarded.

The implication here though is that if you have

```python
@task
def make_a1() -> A1: ...

@task
def use_either(a: typing.Union[A1, A2]): ...
```

This this will fail
```
use_either(a=make_a1())
```
because A1 will match more than one variant.  flytekit itself will not fail I think (right @Future-Outlier?) but we'll never get there because the compiler will fail.

Should we just do exact matches only?  Plus of two examples earlier (case 1 & 2), both will fail mypy type checking.

**Comment by wild-endeavor** [comment]
Thank you @pingsutw and @eapolinario for reminding me... the original issue was an issue because of inheritance.

So the case to make work is

```python
@dataclass
class Parent():
    a: int

@dataclass
class Child(Parent):
    b: int
```
with
```python
def wf():
  c = create_child()  # -> Child
  use_parent(p=c)  # p: Parent
```

The reason we want to make this work is because this works in Python (type checks with mypy) and is common enough we feel that it should work.

So the upshot is, we should do strict subset matching.  If the downstream schema is a subset of the upstream, and the fields that are there are an exact match, then we can.  So in this example, you should be able to pass `Child` to something that expects `Parent` because `Child` has all the fields of `Parent`, and those fields are an exact type match.

Also, `typing.Union[Parent, Child]` is okay to fail because it's ambiguous.


This case should not compile but I think it's okay:
`Child` overwrites a field in `Parent` changing its type, which means it's no longer a strict subset match.
```python
@dataclass
class Parent():
    a: int

@dataclass
class Child(Parent):
    a: Optional[int]
```

This case should compile fine.
```python
@dataclass
class LeafBase():
    leaf: int

@dataclass
class Leaf(LeafBase):
    other: int

@dataclass
class Parent:
    l: LeafBase
```
A task that produces `Parent(l=Leaf(leaf=5, other=3))` should be able to be passed to a downstream task that takes `Parent`.

**Comment by mao3267** [comment]
FYI, originally we tried to use the LGPL package [json_schema_compare](https://pkg.go.dev/gitlab.com/yvesf/json-schema-compare) to do the subset check. Even though feeling pretty sure that the license is compatible, we just not using now without any other specific reasons.

**Comment by Future-Outlier** [comment]
We can create a new workflow version to do this, the upgrade scenario will not fail, I tested it with @pingsutw.
@wild-endeavor also agrees that this is a low priority.
Let's close this and work on other hight priority first.

**Comment by Future-Outlier** [comment]
Sorry I closed the wrong PR... XD

**Review by Future-Outlier — COMMENTED** [review]
just synced with @mao3267 

- work

1. same
2. big -> small
3. child -> parent

- not work

1. different attribute
class A:
	a: int

class B:
	a: int

A -> B

2. parent -> child

**Review by Future-Outlier — COMMENTED** [review]
Just synced with @mao3267 , here is my summary.
1. if we support `child -> parent` case,  we can't support 2 dataclass ambiguous error. 
example:
```python
@dataclass
class A:
	a: int

@dataclass
class B:
	a: int
	
# A -> B can pass and work
```

this is due to json schema's restriction.
we should use MRO to solve this or find other protocol.

File: flytepropeller/pkg/compiler/validators/typing.go [hunk 2]
```diff
@@ -1,9 +1,14 @@
...
 
 import (
+	"bytes"
+	"encoding/json"
 	"strings"
 
 	structpb "github.com/golang/protobuf/ptypes/struct"
+	"github.com/santhosh-tekuri/jsonschema"
+	"github.com/wI2L/jsondiff"
+	jscmp "gitlab.com/yvesf/json-schema-compare"
```
Comment (Future-Outlier) [inline flytepropeller/pkg/compiler/validators/typing.go:11]: `santhosh-tekuri/jsonschema"` uses Apache License.
`wI2L/jsondiff` uses MIT License.
`yvesf/json-schema-compare` uses LGPL License.

**Review by Future-Outlier — COMMENTED** [review]
Just tested all examples you provided, and even do some additional test.
not gonna lie, this is AMAZING.

image

**Review by Future-Outlier — APPROVED** [review]
LGTM, really enjoy to work with you.

**Comment by mao3267** [comment]
> Would it be possible to document as proposed here how type transformers need to provide the schema?

No problem! Will update the documents later.
